module/zfs/arc.c

```diff
@@ -8700,6 +9056,793 @@ l2arc_stop(void)
+	ARCSTAT_INCR(arcstat_l2_psize, asize);
+
+	mutex_enter(&dev->l2ad_mtx);
+	list_insert_tail(&dev->l2ad_buflist, hdr);
+	(void) zfs_refcount_add_many(&dev->l2ad_alloc, asize, hdr);
+	mutex_exit(&dev->l2ad_mtx);
+
+	exists = buf_hash_insert(hdr, &hash_lock);
+	if (exists) {
+		/* Buffer was already cached, no need to restore it. */
+		arc_hdr_destroy(hdr);
+		ARCSTAT_BUMP(arcstat_l2_rebuild_bufs_precached);
+	}
+
+	mutex_exit(hash_lock);
+}
+
+/*
+ * Starts an asynchronous read IO to read a log block. This is used in log
+ * block reconstruction to start reading the next block before we are done
+ * decoding and reconstructing the current block, to keep the l2arc device
+ * nice and hot with read IO to process.
+ * The returned zio will contain a newly allocated memory buffers for the IO
+ * data which should then be freed by the caller once the zio is no longer
+ * needed (i.e. due to it having completed). If you wish to abort this
+ * zio, you should do so using l2arc_log_blk_fetch_abort, which takes
+ * care of disposing of the allocated buffers correctly.
+ */
+static zio_t *
+l2arc_log_blk_fetch(vdev_t *vd, const l2arc_log_blkptr_t *lbp,
+    uint8_t *lb_buf)
+{
+	uint32_t		psize;
+	zio_t			*pio;
+	l2arc_read_callback_t	*cb;
+
+	psize = BLKPROP_GET_PSIZE((lbp)->lbp_prop);
+	ASSERT(psize <= sizeof (l2arc_log_blk_phys_t));
+	cb = kmem_alloc(sizeof (l2arc_read_callback_t), KM_SLEEP);
+	cb->l2rcb_abd = abd_get_from_buf(lb_buf, psize);
+	pio = zio_root(vd->vdev_spa, l2arc_blk_fetch_done, cb,
+	    ZIO_FLAG_DONT_CACHE | ZIO_FLAG_CANFAIL | ZIO_FLAG_DONT_PROPAGATE |
+	    ZIO_FLAG_DONT_RETRY);
+	(void) zio_nowait(zio_read_phys(pio, vd, lbp->lbp_daddr, psize,
+	    cb->l2rcb_abd, ZIO_CHECKSUM_OFF, NULL, NULL,
+	    ZIO_PRIORITY_ASYNC_READ, ZIO_FLAG_DONT_CACHE | ZIO_FLAG_CANFAIL |
+	    ZIO_FLAG_DONT_PROPAGATE | ZIO_FLAG_DONT_RETRY, B_FALSE));
+
+	return (pio);
+}
+
+/*
+ * Aborts a zio returned from l2arc_log_blk_fetch and frees the data
+ * buffers allocated for it.
+ */
+static void
+l2arc_log_blk_fetch_abort(zio_t *zio)
+{
+	(void) zio_wait(zio);
+}
+
+/*
+ * Creates a zio to update the device header on an l2arc device. The zio is
+ * initiated as a child of `pio'.
+ */
+static void
+l2arc_dev_hdr_update(l2arc_dev_t *dev, zio_t *pio,
+    l2arc_write_callback_t *cb)
+{
+	zio_t			*wzio;
+	l2arc_dev_hdr_phys_t	*hdr = dev->l2ad_dev_hdr;
+	const uint64_t		hdr_asize = dev->l2ad_dev_hdr_asize;
+
+	hdr->dh_magic = L2ARC_DEV_HDR_MAGIC;
+	hdr->dh_version = L2ARC_PERSISTENT_VERSION;
+	hdr->dh_spa_guid = spa_guid(dev->l2ad_vdev->vdev_spa);
+	hdr->dh_alloc_space = zfs_refcount_count(&dev->l2ad_alloc);
+	hdr->dh_flags = 0;
+	if (dev->l2ad_first)
+		hdr->dh_flags |= L2ARC_DEV_HDR_EVICT_FIRST;
+
+	/* checksum operation goes last */
+	l2arc_dev_hdr_checksum(hdr, &hdr->dh_self_cksum);
+	cb->l2wcb_abd = abd_get_from_buf(hdr, hdr_asize);
+	wzio = zio_write_phys(pio, dev->l2ad_vdev, VDEV_LABEL_START_SIZE,
+	    hdr_asize, cb->l2wcb_abd, ZIO_CHECKSUM_OFF,
```

Comment:
Since this is a *_phys IO you should be able to set `ZIO_CHECKSUM_LABEL` here and and in `l2arc_dev_hdr_read()`.  This will request that the I/O pipeline embed a checksum, `zio_ect_t`, at the end of the block when it's written.  The pipeline then can verify the checksum on read for you, so you'll get an ECKSUM error from the `zio_read_phys()` and not have to manually handle the checksum validation.  Which is convenient!, and tools like `zdb` will be aware of it.

You could rely on the `zec->zec_magic` for byte swapping if you like, but I think keeping the `dh_magic` is a good idea to make these blocks easily identifiable.  Here's how the `vdev_phys_t` label blocks are defined.

```
typedef struct vdev_phys {
        char            vp_nvlist[VDEV_PHYS_SIZE - sizeof (zio_eck_t)];
        zio_eck_t       vp_zbt;
} vdev_phys_t;
```

The same _could_ be done for the `l2arc_log_blk_phys_t` if we left room at the end then they would be checksummed as well.

Reply:
Nice suggestion. I gave it a minimal try with:
```
--- a/include/sys/arc_impl.h
+++ b/include/sys/arc_impl.h
@@ -221,7 +221,8 @@ typedef struct l2arc_dev_hdr_phys {
         */
        l2arc_log_blkptr_t      dh_start_lbps[2];
        uint64_t                dh_log_blk_ent; /* entries per log blk */
-       uint64_t                dh_pad[42];     /* pad to 512 bytes */
+       uint64_t                dh_pad[37];     /* pad to 512 bytes */
+       zio_eck_t               dh_tail;
 } l2arc_dev_hdr_phys_t;
 CTASSERT_GLOBAL(sizeof (l2arc_dev_hdr_phys_t) == SPA_MINBLOCKSIZE);

diff --git a/module/zfs/arc.c b/module/zfs/arc.c
index 6e7008430..d042ed6c0 100644
--- a/module/zfs/arc.c
+++ b/module/zfs/arc.c
@@ -9285,7 +9285,7 @@ l2arc_dev_hdr_read(l2arc_dev_t *dev)

        err = zio_wait(zio_read_phys(NULL, dev->l2ad_vdev,
            VDEV_LABEL_START_SIZE, hdr_asize, abd,
-           ZIO_CHECKSUM_OFF, NULL, NULL, ZIO_PRIORITY_ASYNC_READ,
+           ZIO_CHECKSUM_LABEL, NULL, NULL, ZIO_PRIORITY_ASYNC_READ,
            ZIO_FLAG_DONT_CACHE | ZIO_FLAG_CANFAIL |
            ZIO_FLAG_DONT_PROPAGATE | ZIO_FLAG_DONT_RETRY, B_FALSE));

```

But I get panics:
```
[Jan 9 23:19] BUG: kernel NULL pointer dereference, address: 0000000000000448
[  +0.000604] #PF: supervisor read access in kernel mode
[  +0.000444] #PF: error_code(0x0000) - not-present page
[  +0.000433] PGD 0 P4D 0
[  +0.000222] Oops: 0000 [#1] SMP PTI
[  +0.000299] CPU: 3 PID: 19901 Comm: z_rd_int Tainted: P           OE     5.4.0-1-vfio-debug-00015-g7776c0538c53 #1
[  +0.000928] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 0.0.0 02/06/2015
[  +0.000786] RIP: 0010:zfs_ereport_start_checksum+0x15a/0x200 [zfs]
[  +0.000520] Code: 74 73 4c 8d ab 18 0d 00 00 4c 8d b3 40 0d 00 00 4c 89 ef e8 08 a2 49 e4 4c 89 f7 e8 40 f2 ff ff 48 8b 85 68 01 00 00 4c 89 f7 <48> 8b 80 48 04 00 00 49 89 04 24 48 8b 85 68 01 00 00 48 8d ab 38
[  +0.001543] RSP: 0018:ffffb68e00ef7cd8 EFLAGS: 00010246
[  +0.000464] RAX: 0000000000000000 RBX: ffff901279004000 RCX: 0000000000000000
[  +0.000594] RDX: ffff9012749a9e80 RSI: 0000000000000000 RDI: ffff901279004d40
[  +0.000595] RBP: ffff9012772879e0 R08: ffff9012749a9e80 R09: 0000000000000020
[  +0.000594] R10: 0000000000000000 R11: 0000000000000200 R12: ffff901276ecf240
[  +0.000594] R13: ffff901279004d18 R14: ffff901279004d40 R15: ffffb68e00ef7d38
[  +0.000598] FS:  0000000000000000(0000) GS:ffff90127bb80000(0000) knlGS:0000000000000000
[  +0.000672] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  +0.000481] CR2: 0000000000000448 CR3: 000000017996e005 CR4: 0000000000360ee0
[  +0.000595] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  +0.000595] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[  +0.000593] Call Trace:
[  +0.000277]  zio_checksum_verify+0x164/0x210 [zfs]
[  +0.000454]  zio_execute+0xe6/0x220 [zfs]
[  +0.000348]  taskq_thread+0x24f/0x4b0 [spl]
[  +0.000361]  ? __switch_to_asm+0x34/0x70
[  +0.000335]  ? wake_up_q+0x60/0x60
[  +0.000342]  ? zio_execute_stack_check.constprop.0+0x10/0x10 [zfs]
[  +0.000518]  kthread+0xfb/0x130
[  +0.000272]  ? sigfillset.isra.0+0x10/0x10 [spl]
[  +0.000390]  ? kthread_park+0x90/0x90
[  +0.000311]  ret_from_fork+0x35/0x40
[  +0.000304] Modules linked in: zfs(POE) icp(POE) zlua(POE) zcommon(POE) zunicode(POE) znvpair(POE) zavl(POE) spl(OE) dm_mod loop cfg80211 mousedev intel_rapl_msr intel_rapl_common rfkill kvm_intel 8021q garp mrp stp llc irqbypass crct1
0dif_pclmul crc32_pclmul ghash_clmulni_intel nls_iso8859_1 nls_cp437 aesni_intel vfat crypto_simd fat cryptd glue_helper psmouse intel_agp iTCO_wdt intel_gtt iTCO_vendor_support i2c_i801 agpgart pcspkr input_leds lpc_ich qemu_fw_cfg evde
v mac_hid ip_tables x_tables ext4 crc32c_generic crc16 mbcache jbd2 ahci libahci libata serio_raw xhci_pci atkbd xhci_hcd libps2 crc32c_intel i8042 serio sd_mod virtio_scsi scsi_mod virtio_net virtio_rng rng_core net_failover failover vi
rtio_pci virtio_blk virtio_ring virtio [last unloaded: spl]
[  +0.005617] CR2: 0000000000000448
[  +0.000289] ---[ end trace e528ce4ee8df14f1 ]---
[  +0.000437] RIP: 0010:zfs_ereport_start_checksum+0x15a/0x200 [zfs]
[  +0.000522] Code: 74 73 4c 8d ab 18 0d 00 00 4c 8d b3 40 0d 00 00 4c 89 ef e8 08 a2 49 e4 4c 89 f7 e8 40 f2 ff ff 48 8b 85 68 01 00 00 4c 89 f7 <48> 8b 80 48 04 00 00 49 89 04 24 48 8b 85 68 01 00 00 48 8d ab 38
[  +0.001547] RSP: 0018:ffffb68e00ef7cd8 EFLAGS: 00010246
[  +0.000440] RAX: 0000000000000000 RBX: ffff901279004000 RCX: 0000000000000000
[  +0.000598] RDX: ffff9012749a9e80 RSI: 0000000000000000 RDI: ffff901279004d40
[  +0.000596] RBP: ffff9012772879e0 R08: ffff9012749a9e80 R09: 0000000000000020
[  +0.000595] R10: 0000000000000000 R11: 0000000000000200 R12: ffff901276ecf240
[  +0.000595] R13: ffff901279004d18 R14: ffff901279004d40 R15: ffffb68e00ef7d38
[  +0.000596] FS:  0000000000000000(0000) GS:ffff90127bb80000(0000) knlGS:0000000000000000
[  +0.000673] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  +0.000482] CR2: 0000000000000448 CR3: 000000017996e005 CR4: 0000000000360ee0
[  +0.000596] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[  +0.000595] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
```

Reply:
That `NULL` pointer dereference is eluding me. I noticed that whenever `vdev_label_read()` is called `ZIO_FLAG_SPECULATIVE` is set which causes `zio_checksum_verify()` to bypass `zfs_ereport_start_checksum()`. Do we need that flag here?

Reply:
I think the culprit is `&zb->io_bookmark`.

Reply:
I got it working. I think we need it in case the device header has not been initialized before.

Reply:
You beat me to it.  Yes, I believe we do want `ZIO_FLAG_SPECULATIVE` set for these phys IOs to bypass `zfs_ereport_start_checksum`.  Sorry, I didn't remember that detail when originally commenting.

Reply:
I will give a try to do the same for `l2arc_log_blk_phys_t`.

Reply:
I gave it a try, but we compress before issuing `zio_write_phys()` using `zio_compress_data()`, and that seems to mess up checksumming.

Reply:
I think having compression of log blocks is definitely beneficial for smaller L2ARC devices.